player.canAfford reworked

sienmich opened the pull request.

kberg commented on src/Player.ts:
```diff
@@ -1578,62 +1578,47 @@ export class Player implements ISerializable<SerializedPlayer> {
-      }
-    }
+  public canAffordCard(card: IProjectCard, reserveUnits : Partial<Units> = Units.EMPTY): boolean {
+    return this.canAffordOptions(this.getCardCost(card), {
```
What does canAffordOptions mean?

kberg commented on src/Units.ts:
```diff
@@ -71,6 +71,17 @@ export namespace Units {
     };
   }
 
+  export function sum(a: Units, b: Units): Units {
```
These methods have a verb for their names, so "add"?

kberg commented on src/Player.ts:
```diff
@@ -1578,62 +1578,47 @@ export class Player implements ISerializable<SerializedPlayer> {
-      maxPay += dirigibles.resourceCount * 3;
-    }
+    // Return those removed units:
+    Units.deductUnits(Units.negative(reserveUnits), this);
```
I'd like to avoid deducting negative units to result in positive units.

kberg commented on src/Player.ts:
```diff
@@ -1578,62 +1578,47 @@ export class Player implements ISerializable<SerializedPlayer> {
-    const psychrophiles = this.playedCards.find(
-      (playedCard) => playedCard.name === CardName.PSYCHROPHILES);
+    // Remove reserveUnits for a moment, to check if we can still afford the card:
+    Units.deductUnits(reserveUnits, this);
```
What I'd like to head towards at some point is that a player's units are a Units object themselves, which would make this a lot easier. And once you start doing this kind of pre-emptive check, we'll need lots of tests to ensure Player's units are untouched when canPlay is false. So, while this is convenient, it's asking for subtle bugs.

kberg reviewed (commented):
What I'm not clear about with this PR is what your goal is. canAfford should eventually get the parameterization treatment, for sure, but I'm not sure what you're aining for  here.

sienmich commented:
@kberg, in fact there was too many things being changed at once. So I changed that, now there are no new functions or changed signatures, I just simplified their meat - counting maximal affordable cost.

kberg commented on src/Player.ts:
```diff
@@ -1604,54 +1605,22 @@ export class Player implements ISerializable<SerializedPlayer> {
   public canAfford(cost: number, canUseSteel: boolean = false, canUseTitanium: boolean = false, canUseFloaters: boolean = false, canUseMicrobes : boolean = false): boolean {
-    let extraResource: number = 0;
-    if (canUseFloaters !== undefined && canUseFloaters) {
-      extraResource += this.getFloatersCanSpend() * 3;
```
How have you verified removing this block doesn't cause breaks?

sienmich replied:
Are you asking why is this working or are you asking to include new tests? :)

I didn't remove this block, I just included it in the sum below (`(canUseFloaters ? this.getFloatersCanSpend() * 3 : 0) +
`)

kberg replied:
I'm asking the former. How do you know it still works. But I'll look, thanks.

kberg reviewed (commented):
This is definitely much simpler!

kberg reviewed (approved):
This seems okay to me. Adding a test to Diriglbles.spec.ts that matches Psychophiles.spec.ts would close the loop, I think. But OK for now.

sienmich commented:
Ok, I'll add `Diriglbles.spec.ts` with next changes
